From 6d1fead62f69959f6c30f16b72623781ba2dd877 Mon Sep 17 00:00:00 2001
From: Naveen <navtat@amazon.com>
Date: Tue, 4 Oct 2022 20:26:11 +0000
Subject: [PATCH] Create CMakeLists patch

Signed-off-by: Naveen <navtat@amazon.com>
---
 jni/CMakeLists.txt | 58 ++++++++--------------------------------------
 1 file changed, 10 insertions(+), 48 deletions(-)

diff --git a/jni/CMakeLists.txt b/jni/CMakeLists.txt
index c6b2f5f..ba9fccd 100644
--- a/jni/CMakeLists.txt
+++ b/jni/CMakeLists.txt
@@ -16,6 +16,7 @@ set(TARGET_LIBS "")  # Libs to be installed
 
 set(CMAKE_CXX_STANDARD 11)
 set(CMAKE_CXX_STANDARD_REQUIRED True)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fpermissive")
 
 option(CONFIG_FAISS "Configure faiss library build when this is on")
 option(CONFIG_NMSLIB "Configure nmslib library build when this is on")
@@ -35,6 +36,11 @@ if (${CMAKE_SYSTEM_NAME} STREQUAL Darwin)
 elseif(${CMAKE_SYSTEM_NAME} STREQUAL Linux)
     set(JVM_OS_TYPE linux)
     set(LIB_EXT .so)
+elseif(${CMAKE_SYSTEM_NAME} STREQUAL Windows)
+    set(JVM_OS_TYPE win32)
+    set(LIB_EXT .dll)
+    set(CMAKE_SHARED_LIBRARY_PREFIX "")
+    set(CMAKE_STATIC_LIBRARY_PREFIX "")
 else()
     message(FATAL_ERROR "Unable to run on system: ${CMAKE_SYSTEM_NAME}")
 endif()
@@ -57,7 +63,7 @@ add_library(${TARGET_LIB_COMMON} SHARED ${CMAKE_CURRENT_SOURCE_DIR}/src/jni_util
 target_include_directories(${TARGET_LIB_COMMON} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/include $ENV{JAVA_HOME}/include $ENV{JAVA_HOME}/include/${JVM_OS_TYPE})
 set_target_properties(${TARGET_LIB_COMMON} PROPERTIES SUFFIX ${LIB_EXT})
 set_target_properties(${TARGET_LIB_COMMON} PROPERTIES POSITION_INDEPENDENT_CODE ON)
-set_target_properties(${TARGET_LIB_COMMON} PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/release)
+set_target_properties(${TARGET_LIB_COMMON} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/release)
 list(APPEND TARGET_LIBS ${TARGET_LIB_COMMON})
 # ----------------------------------------------------------------------------
 
@@ -79,7 +85,7 @@ if (${CONFIG_NMSLIB} STREQUAL ON OR ${CONFIG_ALL} STREQUAL ON OR ${CONFIG_TEST}
     target_include_directories(${TARGET_LIB_NMSLIB} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/include $ENV{JAVA_HOME}/include $ENV{JAVA_HOME}/include/${JVM_OS_TYPE} ${CMAKE_CURRENT_SOURCE_DIR}/external/nmslib/similarity_search/include)
     set_target_properties(${TARGET_LIB_NMSLIB} PROPERTIES SUFFIX ${LIB_EXT})
     set_target_properties(${TARGET_LIB_NMSLIB} PROPERTIES POSITION_INDEPENDENT_CODE ON)
-    set_target_properties(${TARGET_LIB_NMSLIB} PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/release)
+    set_target_properties(${TARGET_LIB_NMSLIB} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/release)
 
     list(APPEND TARGET_LIBS ${TARGET_LIB_NMSLIB})
 endif ()
@@ -130,7 +136,7 @@ if (${CONFIG_FAISS} STREQUAL ON OR ${CONFIG_ALL} STREQUAL ON OR ${CONFIG_TEST} S
     target_include_directories(${TARGET_LIB_FAISS} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/include $ENV{JAVA_HOME}/include $ENV{JAVA_HOME}/include/${JVM_OS_TYPE} ${CMAKE_CURRENT_SOURCE_DIR}/external/faiss)
     set_target_properties(${TARGET_LIB_FAISS} PROPERTIES SUFFIX ${LIB_EXT})
     set_target_properties(${TARGET_LIB_FAISS} PROPERTIES POSITION_INDEPENDENT_CODE ON)
-    set_target_properties(${TARGET_LIB_FAISS} PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/release)
+    set_target_properties(${TARGET_LIB_FAISS} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/release)
 
     list(APPEND TARGET_LIBS ${TARGET_LIB_FAISS})
 endif ()
@@ -138,51 +144,7 @@ endif ()
 # ---------------------------------------------------------------------------
 
 # --------------------------------- TESTS -----------------------------------
-if (${CONFIG_ALL} STREQUAL ON OR ${CONFIG_TEST} STREQUAL ON)
-    # Reference - https://crascit.com/2015/07/25/cmake-gtest/
-    configure_file(CMakeLists.txt.in googletest-download/CMakeLists.txt)
-    execute_process(COMMAND "${CMAKE_COMMAND}" -G "${CMAKE_GENERATOR}" .
-            WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/googletest-download"
-            )
-    execute_process(COMMAND "${CMAKE_COMMAND}" --build .
-            WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/googletest-download"
-            )
-    set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
-
-    add_subdirectory("${CMAKE_BINARY_DIR}/googletest-src"
-            "${CMAKE_BINARY_DIR}/googletest-build" EXCLUDE_FROM_ALL
-            )
-    add_executable(
-            jni_test
-            tests/faiss_wrapper_test.cpp
-            tests/nmslib_wrapper_test.cpp
-            tests/test_util.cpp)
-
-    target_link_libraries(
-            jni_test
-            gtest_main
-            gmock_main
-            faiss
-            NonMetricSpaceLib
-            OpenMP::OpenMP_CXX
-            ${TARGET_LIB_FAISS}
-            ${TARGET_LIB_NMSLIB}
-            ${TARGET_LIB_COMMON}
-    )
-
-    target_include_directories(jni_test PRIVATE
-            ${CMAKE_CURRENT_SOURCE_DIR}/tests
-            ${CMAKE_CURRENT_SOURCE_DIR}/include
-            $ENV{JAVA_HOME}/include
-            $ENV{JAVA_HOME}/include/${JVM_OS_TYPE}
-            ${CMAKE_CURRENT_SOURCE_DIR}/external/faiss
-            ${CMAKE_CURRENT_SOURCE_DIR}/external/nmslib/similarity_search/include
-            ${gtest_SOURCE_DIR}/include
-            ${gmock_SOURCE_DIR}/include)
-
-
-    set_target_properties(jni_test PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/bin)
-endif ()
+
 # ---------------------------------------------------------------------------
 
 # -------------------------------- INSTALL ----------------------------------
-- 
2.37.3.windows.1

